A43-sdk-authorization.md

```diff
@@ -0,0 +1,713 @@
+  									  "- Presence match: \"*\" will match when the"
+  									  "  value is not empty.",
+  									"type": "array",
+  									"items": {
+  										"type": "string"
+  									}
+  								}
+  							},
+  							"required": ["key", "values"]
+  						}
+  					}
+  				}
+  			}
+  		},
+  		"required": ["name"]
+  	}
+  },
+  "description": "AuthorizationPolicy defines which principals are permitted to"
+    "access which resource. Resources are RPC methods scoped by services.",
+  "type": "object",
+  "properties": {
+  	"name": {
+  		"description": "The name of an authorization policy. It is mainly for"
+  		  "monitoring and error message generation.",
+  		"type": "string"
+  	},
+  	"deny_rules": {
+  		"description": "List of deny rules to match. If a request matches any of the"
+  		  "deny rules, then it will be denied. If none of the deny rules matches or"
+  		  "there are no deny rules, the allow rules will be evaluated.",
+  		"type": "array",
+  		"items": {
+  			"$ref": "#/definitions/rule"
+  		}
+  	},
+  	"allow_rules": {
+  		"description": "List of allow rules to match. The allow rules will only be"
+  		  "evaluated after the deny rules. If a request matches any of the allow"
+  		  "rules, then it will allowed. If none of the allow rules matches, it will"
+  		  "be denied.",
+  		"type": "array",
+  		"items": {
+  			"$ref": "#/definitions/rule"
+  		}
+  	}
+  },
+  "required": ["name", "allow_rules"]
+}
+```
+
+#### Details
+
+gRPC SDK authorization policy has a list of *allow rules* and a list of *deny rules*.
+The following sequence is followed to make an authorization decision -
+1. Check for a match in *deny rules*, if a match is found, the request will be denied.
+   If no match is found in deny rules, or there are no deny rules execute next step.
+2. Check for a match in *allow rules*, if a match is found, the request will be allowed.
+   Note that allow rules is a required field here, if not present, policy is invalid.
+3. If no match is found in *allow rules*, we deny the request.
+
+Each *rule* has the following semantics -
+1. Each *rule* has a *name*, a *source* and a *request* to match against. For a *rule*
+   to match, both *source* and *request* must match.
+   - If both *source* and *request* are empty, the *rule* always matches. It is a
+     wildcard that can be used in a deny list to deny all RPCs or in an allow list to
+     accept all RPCs except those matching in the deny list.
+   - If only *source* is empty, we evaluate against the *request* fields and apply
+     that to any user (See example below). Similarly, if only *request* is empty, we
+     evaluate against the *source* fields and apply that to any action.
+2. Each *source* could contain a list of *principals*. The *principals* are ORed
+   together, i.e. it matches if one of them matches.
+   Sequence of steps to evaluate each principal from config -
+   1. If TLS is not used, matching fails.
```

Comment:
You mean "if mTLS is not used..." . Mere TLS doesn't have the client identity available to the server

Reply:
It can be mTLS or TLS. We could have a policy to allow authenticated users to perform some action.

Reply:
But TLS (which is not mTLS) does not provide client authentication. So with TLS there cannot be "authenticated clients". On line 214 you say "if principal matches URI SANs from the client certificate". There is no client certificate in case of a TLS connection which is what I was trying to say.

Reply:
That is a good point. 
Referring to xDS RBAC [proposal](https://github.com/grpc/proposal/blob/master/A41-xds-rbac.md): 
" If principal_name is unset, then Authenticated is said to match if the connection uses TLS; a client certificate is not necessary. " 
Also looking at envoy code, AFAIU its not constrained to mTLS
https://github.com/envoyproxy/envoy/blob/53f95570a7e9d7fa5414a4bc145852524fa017ed/source/extensions/filters/common/rbac/matchers.cc#L206
https://github.com/envoyproxy/envoy/blob/53f95570a7e9d7fa5414a4bc145852524fa017ed/envoy/ssl/connection.h#L79

Referring to my text in the proposal-
"if principal matches URI SANs from the client certificate"
If the peer certificate is not present, it will just skip this matching and move on to the next step
https://github.com/grpc/grpc/blob/master/src/core/lib/security/authorization/matchers.cc#L187

But I do realize that the term authenticated user for TLS seems incorrect, since the client never presents certificates. Let me look into Envoy code and documentation.

Adding @ejona86 for insights.